### ISSUE TITLE
Fully integrating vanilla Emotion as the canonical style system for CSS-in-JS in Gutenberg

### ISSUE THREAD

**Issue by sarayourfriend** [WordPress/gutenberg#30503]
*Note: This issue has been updated significantly, please view the revision history for insight into what's changed!*

This is the next step in fully ingrating Emotion as the canonical style system for Gutenberg.

Recently we made the decision to ditch the new custom style system for the following reasons:

1. It presented too many difficulties in upgrading Emotion to version 11
2. It implemented a custom iframe solution that diverged from Emotion's upstream solution
3. It potentially introduced too many new ways of doing things like themeing and CSS custom property management that would produce too many new public APIs
4. It would introduce a non-trivial maintenance burden on the Gutenberg core contributors

Because @youknowriad was able to figure out how to integrate the `StyleFrame` component using the methods from upstream Emotion for iframe support (thanks Riad!) we're able to fully commit to vanilla Emotion for the Gutenberg style system. This means we do not need a custom wrapper around it.

What are we losing by doing this?

Basically "base styles" which were styles that were being applied to each and every element produced by the style system. Practically this meant the ability to force all components to respect reduced motion automagically. I think we can figure out a solution around this using `injectGlobal` and adding it to the `StyleProvider`, but it's not a blocker, just something to call out.

Okay, so what are the next steps?

- [x] Delete `create-styles` https://github.com/WordPress/gutenberg/pull/31048
- [x] Update `context` to use vanilla emotion https://github.com/WordPress/gutenberg/pull/30877
- [x] Migrate already integrated `ui` components to vanilla emotion
    - [x] Start with `View` which has a cascading effect of needing to handle all components created with `createComponent` and their dependencies, especially variable dependencies. Because we won't be importing from `@wp-g2/styles` anymore, the variables won't exist and we need to port them over to the `COLORS` and `CONFIG` constants from the existing style system in `packages/components/src/utils`. https://github.com/WordPress/gutenberg/pull/31099
    - [x] Continue on with the various other components
        - [x] Delete Button, ButtonGroup (we're ditching these implementations) https://github.com/WordPress/gutenberg/pull/31205
        - [x] Surface https://github.com/WordPress/gutenberg/pull/31238
        - [x] Scrollable https://github.com/WordPress/gutenberg/pull/31207
        - [x] Shortcut https://github.com/WordPress/gutenberg/pull/31229
        - [x] Portal, Popover, Card https://github.com/WordPress/gutenberg/pull/31209
        - [x] Tooltip https://github.com/WordPress/gutenberg/pull/31232
        - [x] VisuallyHidden https://github.com/WordPress/gutenberg/pull/31210
        - [x] Elevation https://github.com/WordPress/gutenberg/pull/31230
        - [x] Divider https://github.com/WordPress/gutenberg/pull/31212
        - [x] FormGroup, ControlLabel https://github.com/WordPress/gutenberg/pull/31234
        - [x] Flex, VStack, HStack https://github.com/WordPress/gutenberg/pull/31243
    - [x] Turn off FontSizePicker and delete it, reverting to the previous, non-g2 implementation https://github.com/WordPress/gutenberg/pull/31287
    - [x] Remove all remaining wp-g2 imports and remove the dependencies https://github.com/WordPress/gutenberg/pull/31292
- [ ] Finish the FontSizePicker integration/finish migrating the components from https://github.com/WordPress/gutenberg/issues/28399
    - [x] Item/ItemGroup https://github.com/WordPress/gutenberg/pull/30097
    - [x] BaseField https://github.com/WordPress/gutenberg/pull/32250
    - [ ] TextInput https://github.com/WordPress/gutenberg/pull/32381
        - [ ] Move utilities used by TextInput over
            - [x] useControlledValue https://github.com/WordPress/gutenberg/pull/33039
        - [ ] Move TextInput over
    - [ ] Dropdown
    - [ ] Slider (Blocked by TextInput)
    - [ ] SelectDropdown
    - [ ] Re-introduce the FontSizePicker'
- [ ] Add base styles to `View` and ensure all components use the View component for their base rendering
    - [x] Fix shortcut not using `View` https://github.com/WordPress/gutenberg/pull/31555
    - [ ] Add base styles to `View`
    - **Should we actually do this?** `View` isn't the basis of the `styled` components so this will only cover a small set of components that would actually get the base styles. Should we just use `injectGlobal` to apply the base styles to `*` with an `!important` on the reduced motion ones?
- [ ] Upgrade components with their newer, G2 implementations
    - [x] Flex https://github.com/WordPress/gutenberg/pull/31297
    - [x] Text https://github.com/WordPress/gutenberg/pull/31291
    - [x] VisuallyHidden
        - [x] Original PR: https://github.com/WordPress/gutenberg/pull/31244
        - [x] Revert: https://github.com/WordPress/gutenberg/pull/31882
        - [ ] Fix (incoming)
        - Currently we're banking on the Emotion 11 Upgrade (immediately below) fixing the issues discovered in this PR: https://github.com/WordPress/gutenberg/pull/31902. We're hoping that upgrading will fix the bugs we found where the `css` from `emotion` doesn't use the correct emotion cache.
- [x] Upgrade Emotion to version 11 ~https://github.com/WordPress/gutenberg/pull/31476~ https://github.com/WordPress/gutenberg/pull/32216
    - Blocked by an issue in `preactjs/compressed-size-check` https://github.com/preactjs/compressed-size-action/issues/59
- [x] Promote new components from `components/src/ui` into `components/src` and export them as experimental or unstable
    - [x] VStack, HStack https://github.com/WordPress/gutenberg/pull/31300
    - [x] Heading https://github.com/WordPress/gutenberg/pull/31299
    - [x] Divider https://github.com/WordPress/gutenberg/pull/31556
    - [x] View https://github.com/WordPress/gutenberg/pull/31542
    - [x] Grid https://github.com/WordPress/gutenberg/pull/31559
    - [x] Elevation https://github.com/WordPress/gutenberg/pull/31614
- [ ] Migrate new components
    - ~Separator https://github.com/WordPress/gutenberg/pull/31612~ Separator is just a Divider with some default props, in that case... just use Divider.
    - [x] ZStack https://github.com/WordPress/gutenberg/pull/31613
    - [ ] Tag https://github.com/WordPress/gutenberg/pull/32214 (BLOCKED)
        - [x] Add types to Icon https://github.com/WordPress/gutenberg/pull/32219
        - [ ] Add types to Tooltip (Blocked by Popover promotion)
        - [ ] Add types to Button
        - [ ] Finally, migrate Tag
    - [ ] Badge (BLOCKED by Tag, they both depend on the same background color utilties)
    - [x] AspectRatio https://github.com/WordPress/gutenberg/pull/33603 - decided not to be implemented as there is native css support
    - [ ] Image
    - [ ] Avatar #33488
    - [x] SegmentedControl https://github.com/WordPress/gutenberg/pull/31937
        - [ ] Fix the focus ring styles for this component
- [x] Remove all remaining "next" component wrappers https://github.com/WordPress/gutenberg/pull/32205
- [ ] Decide what components should be uplifted and which should be replaced
- [ ] Replace components that need replacing
    - [ ] AccessiblePopover
        - [ ] Integrate AccessiblePopover
            - [x] Replace the `Card` component with the new version from `ui` https://github.com/WordPress/gutenberg/pull/32243
              - [x] Promote `Surface` https://github.com/WordPress/gutenberg/pull/32439
              - [x] Promote `Scrollable` https://github.com/WordPress/gutenberg/pull/32446
              - [x] Refactor folder structure: #32557
              - [x] Add types: #32561
              - [x] Update implementation to g2: #32566
              - [x] Update storybook: #32763
              - [x] Delete old g2 version: #32764
            - [x] Promote Popover https://github.com/WordPress/gutenberg/pull/32197/
        - [ ] Replace its usages
            - [ ] Dropdown
            - [ ] Tooltip
            - [ ] etc...
    - [x] ColorPicker
        - Blocked by several other components needing to be migrated like TextInput
    - [ ] ...
- [ ] Uplift components that need uplifting
    - [ ] Button
    - [ ] Continue making decisions about what components need "uplifting"

## Miscellaneous tasks

- [x] Remove `color` and `config` utility functions and just use the `COLORS` and `CONFIG` objects directly
    - [x] `color` https://github.com/WordPress/gutenberg/pull/31661
    - [x] `config` https://github.com/WordPress/gutenberg/pull/31646
- [x] Remove the implicit `children` prop from `ViewOwnProps` and add `children` to all components that need it explicitly rather than relying on the implicit definition. https://github.com/WordPress/gutenberg/pull/31817
- [x] Remove the `components/src/utils/space` function and just use `ui/utils/space`. https://github.com/WordPress/gutenberg/pull/32852
- [x] Rename `ViewOwnProps` to a clearer name like `PolymorphicComponentProps` https://github.com/WordPress/gutenberg/pull/32053
- [x] Allow components to declare that they are _not_ polymorphic (like Divider for example). https://github.com/WordPress/gutenberg/pull/32796
- [x] Delete unused types from `context/polymorphic-component.ts`

## Component uplifting tasks

- [x] Soft deprecate the various `isPrimary`,`isSecondary` etc. props on `Button` in favor of a single `variant` prop. https://github.com/WordPress/gutenberg/pull/31713
- [ ] Hard deprecate props that were soft deprecated in https://github.com/WordPress/gutenberg/pull/31713 (plus the `isPrimary` key used in the notice component)
- [ ] Extract the `Button` parts of `Button` into a `BaseButton` or `PrimitiveButton` component that does not include the `Tooltip` stuff.

## CSS-in-JS tasks

A list of high priority components to be refactored is tracked in https://github.com/WordPress/gutenberg/issues/35744

- [ ] Convert components to CSS-in-JS
    * Add each component as it is converted to this list here
    - [ ] Button
    - [ ] ButtonGroup
    - [ ] Draggable
    - [ ] CheckboxControl
    * etc...

## Context connecting tasks

A list of high priority components to be refactored is tracked in https://github.com/WordPress/gutenberg/issues/35744

- [ ] Connect each component to the `ui/context`
    * Add each component as it is connected to this list
    - [ ] Button
    - [ ] ButtonGroup
    * etc..

## Documentation tasks

- [ ] Update root `@wordpress/components` README.md to describe the structure of the package including the differences between primitive components (like Text) and composed components (like ColorPicker) 
    - [x] Create a tracking issue for documenting this https://github.com/WordPress/gutenberg/issues/33111
    - [x] Document each component's type
    - [ ] Update each component's README.md with the new information
    - [ ] Update the root README.md with the general description of the component types and their implications

<h2 id="deprecating-emotion-css">Deprecating `@emotion/css`</h2>

- [x] Add an error for `@emotion/css` as a restricted import https://github.com/WordPress/gutenberg/pull/33051
- [ ] Start to remove usages of `@emotion/css` from the codebase and replace with `@emotion/styled` and `@emotion/react`
    - [x] ZStack https://github.com/WordPress/gutenberg/pull/33053
    - [x] Divider https://github.com/WordPress/gutenberg/pull/33054
    - [ ] Spacer https://github.com/WordPress/gutenberg/pull/33055
    - [ ] Surface
    - [ ] Text
    - [ ] Truncate
    - [ ] ControlGroup
    - [ ] ControlLabel
    - [ ] FormGroup
    - [ ] ItemGroup
    - [ ] BaseField
    - [ ] Card and its sub-components
    - [ ] Elevation
    - [ ] Flex

**Comment by gziolo** [comment]
Thank you @sarayourfriend for actively seeking better ways to integrate the new Component System. I think it makes perfect sense to try a bit different approach based on all the experience you collected when working on integration so far. In fact, it's closer to what I imagined in the early phase of planning. I hoped we can switch a few lower-level components to their revised implementation and bring all the necessary machinery to make them work with all the new features the that Component System brings. At that stage (https://github.com/WordPress/gutenberg/issues/27484#issuecomment-739003356), I wasn't aware of the limitation you listed but it only shows that the sooner you start integrating components the more you learn about the potential issues that should be addressed with higher priority. 

I agree that #30509 is the most important blocker to resolve before we can start using some of the components that are already added to the `@wordpress/components/ui` folder 👍🏻

**Comment by jasmussen** [comment]
Question, as the new system is getting integrated, does it make sense to rename `--wp-g2-` prefix to simply `--wp-`, to match existing values?

**Comment by sarayourfriend** [comment]
> Question, as the new system is getting integrated, does it make sense to rename --wp-g2- prefix to simply --wp-, to match existing values?

@jasmussen Yes I think ultimately that should be the goal. For now the new style system, as it is integrated, has the `--wp-experimental` prefix:

https://github.com/WordPress/gutenberg/blob/b53c5e51fdb0db517f71eb80740cf698db344b5b/packages/components/src/ui/create-styles/create-style-system/constants.js#L4

**Comment by jasmussen** [comment]
I appreciate you clarifying that to me, thank you!

**Comment by sarayourfriend** [comment]
**Note!** I've updated the issue to reflect the current plan. Please review when y'all have time @mtias @gziolo @youknowriad. Thanks!

**Comment by gziolo** [comment]
Thank you for the update. It looks like everything is on track. 👍🏻

**Comment by ciampo** [comment]
> Remove color and config utility functions and just use the COLORS and CONFIG objects directly

Opened https://github.com/WordPress/gutenberg/pull/31661 and https://github.com/WordPress/gutenberg/pull/31646

>  Deprecate the various isPrimary,isSecondary etc. props on Button in favor of a single variant prop.

Opened draft PR https://github.com/WordPress/gutenberg/pull/31713

**Comment by sarayourfriend** [comment]
## Migrating a G2 component into Gutenberg.

Helpful example PRs:
* https://github.com/WordPress/gutenberg/pull/31207
* https://github.com/WordPress/gutenberg/pull/31230

The rough outline of the process is as follows
1. Copy the component from the g2 repo into packages/components/src/component-name (use kebab case for the folder)
2. Add the folder to components/tsconfig.json#includes
3. Rename the hook file to `hook.ts` and the component file to `component.ts(x)` and the styles file to `styles.ts`.
4. Update the imports to import `cx` and `css` from emotion.
5. Remove `react` imports and change them to be from `@wordpress/element`
6. Remove the `ui.get` usage and replace it with a combo of the `COLORS` and `CONFIG` objects (you can see an example of `CONFIG` in the Elevation PR I linked. `COLORS` is very similar)
7. Any other functions called off of `ui` that don’t currently have a utility in the Gutenberg repo need to have a utility created.
8. Move any functions imported from `@wp-g2/utils` directly into the Gutenberg repository under the `ui/utils` folder (like just copy them from G2 into GB)
9. Change `@wp-g2/context` imports to import from `ui/context`
10. Fix/add types if they’re missing (they almost certainly are missing, but you can find mostly complete Prop types by searching `ComponentName.d.ts` in the G2 repo). Also be sure to add the `children` type which is implicit in the G2 repository but needs to be explicit in the Gutenberg repository.

Usages of the `ui.$` function can be completely removed. Usages of `createToken` need to be replaced with functions that return CSS based on the variables, similar to the `Surface` PR: https://github.com/WordPress/gutenberg/pull/31238

**Comment by sarayourfriend** [comment]
Regarding base styles, I've added a new comment copied here for ease:

> **Should we actually do this?** `View` isn't the basis of the `styled` components so this will only cover a small set of components that would actually get the base styles. Should we just use `injectGlobal` to apply the base styles to `*` with an `!important` on the reduced motion ones?

What do y'all think?

cc @ciampo @youknowriad

The goal is that we want reduced motion to be respected automagically, but maybe there's no way to do that and we just need to use the `useReducedMotion` hook everywhere 🤔 But I don't know how well that will work in CSS-in-JS as we don't want to depend on runtime changes 🤔 We should be able to apply the media query _and_ potentially override animations if the environment variable is present to disable animations.

Overall just a little stuck thinking about this and curious what other people think.

**Comment by ciampo** [comment]
I may lack some perspective of the high-level strategy for the components package, but it makes sense to me that, whatever solution we may go for, we would be able to affect _all_ components. Having implementation discrepancies in between components would make it quite hard to maintain and implement features effectively.

> `View` isn't the basis of the `styled` components so this will only cover a small set of components that would actually get the base styles

Hopefully not a silly question, but — Can't we make changes so that _all_ components are based on `View`? Would it be too complicated/disruptive?

> [...] But I don't know how well that will work in CSS-in-JS as we don't want to depend on runtime changes 🤔  We should be able to apply the media query and potentially override animations if the environment variable is present to disable animations.

Wouldn't be able to wrap some CSS into a `@media (prefers-reduced-motion) { ... }` query without relying on the `useReducedMotion` hook's dynamic changes?

**Comment by sarayourfriend** [comment]
> Hopefully not a silly question, but — Can't we make changes so that all components are based on View? Would it be too complicated/disruptive?

Not a silly question. All of the new components ultimately render a `View` of some sort, but along with that `View` they may render a `styled` component. I guess we could make the `styled` components all inherit from `View` (``` styled( View )`` ``` for example) but that's not semantic Emotion in any way at all and it would make getting semantic elements out of `styled` more difficult (you'd have to pass the `as` prop rather than just calling off of `styled`).

Previously, when we were using our own wrapper around Emotion, we were able to make all `styled` components ultimately inherit from `View`, but that's not possible without the wrapper around Emotion generating our own custom `styled` object as far as I know.

> Wouldn't be able to wrap some CSS into a `@media (prefers-reduced-motion) { ... }` query without relying on the useReducedMotion hook's dynamic changes?

As far as I understand it, the `FORCE_REDUCED_MOTION` environment variable doesn't affect the `prefers-reduced-motion` query at all, so we always have to check the `FORCED_REDUCED_MOTION` variable in order for it to work with e2e test expectations. Maybe that's an incorrect understanding though, @gziolo @youknowriad can you confirm this one way or the other? Does the env variable effect the media query?

**Comment by annezazu** [comment]
Closing this out as it appears much of this work has been completed or is no longer relevant today (it's been three years since last updated). Happy to reopen if that's incorrect!
